accepted/2025/direct-tool-execution.md

```diff
@@ -0,0 +1,108 @@
+| `--interactive`          | (Optional) Allows the command to stop and wait for user input or action (for example to complete authentication). |
+| `--yes`                  | (Optional) Overrides confirmation prompt with "yes" value.                                 |
+| `-y`                     | (Optional) Overrides confirmation prompt with "yes" value.                                 |
+| `-- [tool arguments]`    | Arguments to pass to the tool being executed.                                                |
+
+**Example usage:**
+
+```sh
+dotnet tool exec dotnetsay --version 1.2.3 -- Hello, world!
+```
+
+Note: The version option would work the same as other dotnet tool commands supporting both an exact version and * wildcard matching
+
+## Detailed Design
+
+- When invoked, `dotnet tool exec` will:
+  1. Check if the specified tool and version are present in the NuGet cache.
+  2. If not present, download and install the tool package to the cache.
+  3. Launch the tool with the provided arguments.
+- The tool is not added to any manifest and is not available globally or locally after execution.
+- The command supports passing all arguments after `--` directly to the tool.
+- Tools executed via `dotnet tool exec` are always installed to the NuGet cache, leveraging existing NuGet cache policies and avoiding the need for a manifest. This simplifies the process and ensures ephemeral tool usage.
+
+## Considerations
+
+- The tool is ephemeral and not discoverable via `dotnet tool list`.
+- Caching behavior follows standard NuGet cache policies.
+
+### Calling by Command vs. Package
+
+There is the potential of supporting calling a tool by its command name rather than the package ID. This approach could improve usability, but it is potentially problematic because .NET tools do not have a standard naming convention for commands, and there are existing protections on nuget.org to guard against spelling errors and package squatting that apply only to package IDs. Further discussion and feedback are needed to determine the best approach for this scenario.
+
+We could support a standard naming convention for the package like "dotnet-<command>" as that's used in dotnet-ef and many other tools but that would open the question of what to do about existing tools.
+
+---
+
+## `dnx` Summary
+
+This proposal introduces a new launcher script, `dnx`, which acts as a thin wrapper for `dotnet tool exec`. The script will be installed as `dnx.cmd` on Windows and as an executable `dnx` (no extension) on non-Windows platforms, and will reside alongside the `dotnet` host executable.
+
+When invoked, `dnx` will forward all arguments to `dotnet tool dnx` which will handle parsing the arguments and passing them through to `dotnet tool exec`, enabling a familiar and concise entry point for running .NET tools directly. The script will not contain any additional logic, ensuring consistency and simplifying maintenance across platforms.
```

Comment:
What the rationale for the extra `dnx` layer?

Reply:
We want to keep the script super simple and hopefully it will never change.  Having the logic in the CLI (and I would put it under `dotnet dnx` instead of `dotnet tool dnx`) leaves the door open for `dnx` to support executing other things, like C# "script" files or projects.  That way the CLI can decide how to interpret `dnx` instead of having to implement behavior in the script.

Reply:
Keeping the script minimal makes sense.

I have been wondering about this:

> `dotnet tool dnx` which will handle parsing the arguments and passing them through to `dotnet tool exec

This suggests that `dotnet dnx` will parse the arguments, does something with them (what is that exactly?), creates a new command line to pass to `dotnet tool exec`, `dotnet tool exec` parses the arguments again and executes the tool. Can `dotnet dnx` just do the thing?

Reply:
When I was originally writing the doc, I had the idea of dnx detecting if it was a tool, an exe, a dll, or a .cs file and running different commands. I think the parsing text was left over from that as we decided to move that to a non-goal to scope and simplify the feature. I can't think of anything that dnx should do that tool exec couldn't.

I'll clarify that we won't actually create a new command line for dotnet tool exec but just be using the same command logic internally. I definitely don't want to spin up an extra dotnet process.